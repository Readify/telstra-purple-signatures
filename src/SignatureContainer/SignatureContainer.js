import React from 'react';
import { entries } from 'lodash/object';
import { compose, mapProps, withProps } from 'recompose';

import constants from '../constants';
import './SignatureContainer.scss';
import { PurpleSignatureContainer } from './presentational-components';
import { createButtons } from './createSignatureButtons';
import { stripObject } from '../util';

const assignPlaceholders = (props, placeholders) => {
  const isBlank = a => a === '' || a === null || a === undefined;
  return entries(props).reduce((result, item) => {
    const [key, value] = item;
    return Object.assign(result, {
      [key]: isBlank(value) ? placeholders[key] : value
    });
  }, {});
};

export const containerChooser = props => ({
  Container: PurpleSignatureContainer
});

export const createContainer = props => {
  const {
    name,
    pronoun,
    title,
    qualifications,
    twitter,
    sigType,
    mobile,
    email,
    supportHotline,
    supportEmail,
    Container,
    brandLogo,
    brandName,
    brandLink,
    brandLinkName
  } = props;

  const placeholders = constants.purple.placeholders;
  const brandInfo = constants.purple.brandInfo;

  const signatureProps = stripObject({
    title,
    qualifications,
    twitter,
    phone: null,
    ...assignPlaceholders(
      stripObject({
        name,
        pronoun,
        mobile,
        email,
        sigType,
        supportHotline,
        supportEmail
      }),
      placeholders
    ),
    ...assignPlaceholders(
      { brandLogo, brandName, brandLink, brandLinkName },
      brandInfo
    )
  });

  const buttons = createButtons(signatureProps, placeholders, false);

  return {
    containerProps: {
      signatureProps,
      ...buttons
    },
    Container
  };
};

const SignatureContainer = ({ containerProps, Container }) => (
  <Container {...containerProps} />
);

export default compose(
  withProps(containerChooser),
  mapProps(createContainer)
)(SignatureContainer);
