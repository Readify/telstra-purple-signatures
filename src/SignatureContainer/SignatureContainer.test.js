import React from 'react';
import { shallow } from 'enzyme';

import SignatureContainer, { createContainer } from './SignatureContainer';

describe('SignatureContainer', () => {
  it('Renders', () => {
    const props = {
      email: 'your.name@telstra.purple.net',
      sigType: 'purple',
      mobile: '+61 111 222 333',
      name: 'your name',
      pronoun: 'he/him | she/her | they/them',
      title: 'Job Title',
      twitter: '@myTwitter'
    };

    const tree = shallow(<SignatureContainer {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('Renders with support fields', () => {
    const props = {
      email: 'your.name@telstra.purple.net',
      sigType: 'support',
      mobile: '+61 111 222 333',
      name: 'your name',
      pronoun: 'he/him | she/her | they/them',
      title: 'Job Title',
      twitter: '@myTwitter'
    };

    const tree = shallow(<SignatureContainer {...props} />);
    expect(tree).toMatchSnapshot();
  });
});

describe('createContainer', () => {
  it('returns the right props', () => {
    const props = {
      email: 'your.name@telstra.purple.net',
      mobile: '+61 111 222 333',
      phone: '+61 123 456 789',
      name: 'your name',
      pronoun: 'he/him | she/her | they/them',
      title: 'Job Title',
      twitter: '@myTwitter'
    };

    expect(createContainer(props)).toMatchSnapshot();
    expect(createContainer({ ...props, sigType: 'purple' })).toMatchSnapshot();
  });
});
