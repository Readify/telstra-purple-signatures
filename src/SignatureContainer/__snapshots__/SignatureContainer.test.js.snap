// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignatureContainer Renders 1`] = `
<mapProps(SignatureContainer)
  Container={[Function]}
  email="your.name@telstra.purple.net"
  mobile="+61 111 222 333"
  name="your name"
  pronoun='he/him | she/her | they/them'
  sigType="purple"
  title="Job Title"
  twitter="@myTwitter"
/>
`;

exports[`SignatureContainer Renders with support fields 1`] = `
<mapProps(SignatureContainer)
  Container={[Function]}
  email="your.name@telstra.purple.net"
  mobile="+61 111 222 333"
  name="your name"
  pronoun='he/him | she/her | they/them'
  sigType="support"
  title="Job Title"
  twitter="@myTwitter"
/>
`;

exports[`createContainer returns the right props 1`] = `
Object {
  "Container": undefined,
  "containerProps": Object {
    "CopyRepliesAndForwardsHtml": <withState(withState(withHandlers(mapProps(Button))))
      classAfter="button is-success"
      classBefore="button is-primary"
      onClickHandler={[Function]}
      textAfter="Copied!"
      textBefore="Copy signature"
    />,
    "CopyRepliesAndForwardsText": <withState(withState(withHandlers(mapProps(Button))))
      classAfter="button is-success"
      classBefore="button"
      onClickHandler={[Function]}
      textAfter="Copied!"
      textBefore="Copy text only"
    />,
    "CopySignatureHtml": <withState(withState(withHandlers(mapProps(Button))))
      classAfter="button is-success"
      classBefore="button is-primary"
      onClickHandler={[Function]}
      textAfter="Copied!"
      textBefore="Copy signature"
    />,
    "CopySignatureText": <withState(withState(withHandlers(mapProps(Button))))
      classAfter="button is-success"
      classBefore="button"
      onClickHandler={[Function]}
      textAfter="Copied!"
      textBefore="Copy text only"
    />,
    "signatureProps": Object {
      "brandLink": "https://purple.telstra.com/",
      "brandLinkName": "purple.telstra.com",
      "brandLogo": Object {
        "alt": "Telstra Purple",
        "link": "https://readifysignatures.blob.core.windows.net/images/purple-email-logo.png",
      },
      "brandName": "Telstra Purple",
      "email": "your.name@telstra.purple.net",
      "mobile": "+61 111 222 333",
      "name": "your name",
      "pronoun": "he/him | she/her | they/them",
      "title": "Job Title",
      "twitter": "@myTwitter",
    },
  },
}
`;

exports[`createContainer returns the right props 2`] = `
Object {
  "Container": undefined,
  "containerProps": Object {
    "CopyRepliesAndForwardsHtml": <withState(withState(withHandlers(mapProps(Button))))
      classAfter="button is-success"
      classBefore="button is-primary"
      onClickHandler={[Function]}
      textAfter="Copied!"
      textBefore="Copy signature"
    />,
    "CopyRepliesAndForwardsText": <withState(withState(withHandlers(mapProps(Button))))
      classAfter="button is-success"
      classBefore="button"
      onClickHandler={[Function]}
      textAfter="Copied!"
      textBefore="Copy text only"
    />,
    "CopySignatureHtml": <withState(withState(withHandlers(mapProps(Button))))
      classAfter="button is-success"
      classBefore="button is-primary"
      onClickHandler={[Function]}
      textAfter="Copied!"
      textBefore="Copy signature"
    />,
    "CopySignatureText": <withState(withState(withHandlers(mapProps(Button))))
      classAfter="button is-success"
      classBefore="button"
      onClickHandler={[Function]}
      textAfter="Copied!"
      textBefore="Copy text only"
    />,
    "signatureProps": Object {
      "brandLink": "https://purple.telstra.com/",
      "brandLinkName": "purple.telstra.com",
      "brandLogo": Object {
        "alt": "Telstra Purple",
        "link": "https://readifysignatures.blob.core.windows.net/images/purple-email-logo.png",
      },
      "brandName": "Telstra Purple",
      "email": "your.name@telstra.purple.net",
      "mobile": "+61 111 222 333",
      "name": "your name",
      "pronoun": "he/him | she/her | they/them",
      "sigType": "purple",
      "title": "Job Title",
      "twitter": "@myTwitter",
    },
  },
}
`;
